test/test_rpc.py

```diff
@@ -0,0 +1,170 @@
+import sys
+import tempfile
+
+import torch
+import torch.distributed as dist
+import torch.multiprocessing as mp
+
+from common_utils import TestCase, load_tests, run_tests
+from common_utils import NO_MULTIPROCESSING_SPAWN
+
+# load_tests from common_utils is used to automatically filter tests for
+# sharding on sandcastle. This line silences flake warnings
+load_tests = load_tests
+
+if not dist.is_available():
+    print('c10d not available, skipping tests')
+    sys.exit(0)
+
+
+if NO_MULTIPROCESSING_SPAWN:
+    print('spawn not available, skipping tests')
+    sys.exit(0)
+
+class RpcTest(TestCase):
+
+    world_size = 2
+
+    @classmethod
+    def opts(cls, threads=2):
+        opts = dist.ProcessGroupGloo.Options()
+        opts.devices = [dist.ProcessGroupGloo.create_tcp_device(interface="lo")]
+        opts.timeout = 5.0
+        opts.threads = threads
+        return opts
+
+    @classmethod
+    def _init_rpc(cls, rank, filename, world_size):
+        store = dist.FileStore(filename, world_size)
+        dist.init_process_group(
+            backend='gloo', rank=rank, world_size=world_size, store=store)
+        dist.init_rpc('worker%d' % rank)
+
+    @classmethod
+    def _destroy_rpc(cls):
+        dist.destroy_rpc()
+        dist.destroy_process_group(dist.group.WORLD)
+
+    # Why classmethod? multiprocessing cannot pickle TestCase subclass when in
+    # spawn mode. See https://bugs.python.org/issue33884.
+    @classmethod
+    def _test_process(cls, rank, func, filename, c2p, p2c, world_size):
+        RpcTest._init_rpc(rank, filename, world_size)
+
+        result, expected = func(rank, world_size)
+        c2p.put((rank, expected, result))
+        p2c.get()
+
+        RpcTest._destroy_rpc()
+
+    def _test_multiprocess(self, func, world_size):
+        # file store will delete the test file on destruction
+        file = tempfile.NamedTemporaryFile(delete=False)
+        ctx = mp.get_context('spawn')
+        c2p = ctx.Queue(10)
+        p2c = ctx.Queue(10)
+
+        ps = []
+        for i in range(world_size):
+            p = ctx.Process(
+                target=RpcTest._test_process,
+                args=(i, func, file.name, c2p, p2c, world_size))
+            p.start()
+            ps.append(p)
+
+        for _ in range(world_size):
+            rank, expected, result = c2p.get()
+            self.assertEqual(
+                expected,
+                result,
+                (
+                    "Expect rank {} to receive tensor {} but got {}."
+                ).format(rank, expected, result)
+            )
+
+        for _ in range(world_size):
+            p2c.put(0)
+
+        for p in ps:
+            p.join(2)
+
+    @classmethod
+    def _test_add(cls, rank, world_size):
+        n = rank + 1
+        dstRank = n % world_size
+        ret = dist.rpc_sync('worker%d' % dstRank,
+                            'aten::add', torch.ones(n, n), torch.ones(n, n))
```

Comment:
Hi @zdevito `torch.jit._find_builtin(torch.add)` works, but using `torch.jit._find_builtin(torch.ops.aten.add)` leads to the following error. 

```
RuntimeError: Found 13 overloads for operator aten::add! Overloads are not supported from Python.
```

I am a little confused regarding where should we use the global namespace. Wouldn't it be better if users can pass `torch.add` instead of `torch.ops.aten.add`? If so, how do we convert `torch.add` into `torch.ops.aten.add`? It seems `_find_builtin(fn)` only returns `aten::add`, should I manually convert that into `torch::ops::aten::add`?

Reply:
Yeah, let me clarify:

The user writes `torch.add`, internally we realize it is a builtin and resolve it to `aten::add`, in the RPC, we then send `torch.ops.aten.add`. Remotely we recognize that the op is a builtin because it starts with `torch.ops` (all builtins start with this), and we know that it is `aten::add` because all builtins have exactly one namespace and one op name. Doing it this way means we do not have to take the qualified name with what kind of function it is, because all functions are unique in the global namespace. My other comment shows how we can distinguish between the three cases (python function, TorchScript function, builtin)

Reply:
Got it, thanks!!

Reply:
Hi @zdevito, I am thinking about the tradeoff we are making here. IIUC, the idea is to send the full op name (with global namespace) and get rid of the `MessageType` field. The receiver would use the name to figure out the message type. 

Does this mean that the `qualified_name` needs to be a field in `Message` instead of a field in `BuiltinOp` (or `PythonUdfOp`), because we need that information to determine how to deserialize the message? 

On the receiver side, without the message type, it needs to go through a more expensive step to resolve names to message types. Is this overhead worth it to save just a few bytes (1 byte is enough if we optimize it well) in comm?

Besides, what about responses? It doesn't make much sense to include the full op name in the message for `BuiltinRet` or `PythonUdfRet` types. We could ask the caller to remember the type information of the response when sending the request, but wouldn't it be better if the type information comes together with the message itself?

Reply:
@zdevito I agree that it's much better to use `torch.add` instead of `"aten::add"` in the API. In the RPC, as we have to do full name resolving on the caller side (to report errors earlier), is it acceptable that we still use `MessageType` to distinguish builtins, UDF, TorchScript and to avoid resolving the names again on the callee side?

Reply:
Oh, I missed one of your comment below. Would I be correct if I assume that we still keep `MessageType` which contains `SCRIPT_CALL, SCRIPT_RET, PYTHON_CALL, PYTHION_RET`. Then within `SCRIPT_CALL` we use op names to distinguish whether it is builtins or TorchScripts?